src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -0,0 +1,211 @@
+            if (_vsOnlineServices.ConnectedToVSOnline)
+            {
+                return;
+            }
+
+            if (await IsAutomationFunctionAsync() || node.IsFolder)
+            {
+                // Do not display rename Prompt
+                await base.RenameAsync(context, node, value);
+                return;
+            }
+
+            // Get the list of possible actions to execute
+            string oldName = Path.GetFileNameWithoutExtension(_oldFilePath);
+            CodeAnalysis.Project? project = GetCurrentProject();
+            if (project is null)
+            {
+                return;
+            }
+
+            CodeAnalysis.Document? oldDocument = GetDocument(project, _oldFilePath);
+            if (oldDocument is null)
+            {
+                return;
+            }
+
+            var documentRenameResult = await CodeAnalysis.Rename.Renamer.RenameDocumentAsync(oldDocument, _newFileNameWithExtension, oldDocument.Folders);
+
+            bool errorsDetected = false;
+            foreach (var action in documentRenameResult.ApplicableActions)
+            {
+                foreach (var error in action.GetErrors())
+                {
+                    errorsDetected = true;
+                    break;
+                }
+            }
+
+            // Check errors before applying changes
+            if (errorsDetected)
+            {
+                string failureMessage = string.Format(CultureInfo.CurrentCulture, VSResources.CannotApplyRename, oldName);
+                _userNotificationServices.ShowWarning(failureMessage);
+                return;
+            }
+
+            // Rename the file
+            await base.RenameAsync(context, node, value);
+
+            //Check if there are any symbols that need to be renamed
+            if (documentRenameResult.ApplicableActions.IsEmpty)
+                return;
+
+            // Ask if the user wants to rename the symbol
+            bool userWantsToRenameSymbol = await CheckUserConfirmation(oldName);
+            if (!userWantsToRenameSymbol)
+                return;
+
+            _threadingService.RunAndForget(async () =>
+            {
+                // TODO - implement PublishAsync() to sync with LanguageService
+                // https://github.com/dotnet/project-system/issues/3425)
+                // await _languageService.PublishAsync(treeVersion);
+                //
+                // Because HandleRename() is called after Roslyn, we can set this 
+                // flag when node.FilePath == HandlRename(FilePath)
+                await _semLanguegeServiceDone.WaitAsync();
+
+                // Apply actions and notify other VS features
+                CodeAnalysis.Solution currentSolution = GetCurrentProject().Solution;
+                string renameOperationName = string.Format(CultureInfo.CurrentCulture, VSResources.Renaming_Type_from_0_to_1, oldName, value);
```

Comment:
If you want, a very similar string is already provided by roslyn with the `GetDescription()` on the action. 

https://github.com/dotnet/roslyn/blob/a65653a77b2862329052d84816b5fb324f8bc8bd/src/Workspaces/Core/Portable/Rename/Renamer.RenameSymbolDocumentAction.cs#L38

Reply:
https://github.com/dotnet/roslyn/issues/44220